APPLE-179: Add Automation Option to Prepend to Headline

kevinfodness opened the pull request:
## Summary

Augments the Prepend Text option in Automation by allowing users to prepend to the metadata title or the headline or both (by using both options).

## Changelog entries

### Added

- Functionality to prepend text on the headline for publishers.

kevinfodness committed:
WIP: APPLE-179

kevinfodness committed:
APPLE-179 Add new option to prepend text to title component

dlh01 commented on admin/class-automation.php:
```diff
@@ -179,11 +207,11 @@ public static function filter__apple_news_exporter_slug( $slug, $post_id ) {
-		$title = $json['title'] ?? '';
-		$rules = self::get_automation_for_post( $post_id );
+		$rules         = self::get_automation_for_post( $post_id );
+		$json['title'] = self::$original_title;
```
Is there any reason to suspect that `filter__apple_news_exporter_title()` and this method might not run in the "correct" order, such that this static `$original_title` variable references the wrong post?

Since the post ID is available in both places, would it be overkill to make `$original_title` a map of ID > title that could be referenced here?

kevinfodness replied:
They should never run in the wrong order. The exporter title is set up when the exporter object is set up, which happens before any of the logic to build the exported object. However, I like your suggestion about establishing the ID/title map, and will apply that.

dlh01 commented on admin/class-automation.php:
```diff
@@ -170,6 +178,26 @@ public static function filter__apple_news_exporter_slug( $slug, $post_id ) {
+		$rules                = self::get_automation_for_post( $post_id );
+		foreach ( $rules as $rule ) {
+			if ( 'headline.prepend' === ( $rule['field'] ?? '' ) ) {
+				$title = sprintf( '%s %s', $rule['value'] ?? '', self::$original_title );
```
Unconditionally applying the `sprintf()` here in conjunction with the null coalescing to `''` reads a little strangely to me, since it implies that the user would want there to be leading spaces in the title in the event that `value` was missing. I wonder whether there should be a string length check prior to changing `$title`?

kevinfodness replied:
Good point. I'll modify this to look before I leap.

dlh01 commented on admin/class-automation.php:
```diff
@@ -170,6 +178,26 @@ public static function filter__apple_news_exporter_slug( $slug, $post_id ) {
+		$rules                = self::get_automation_for_post( $post_id );
+		foreach ( $rules as $rule ) {
+			if ( 'headline.prepend' === ( $rule['field'] ?? '' ) ) {
+				$title = sprintf( '%s %s', $rule['value'] ?? '', self::$original_title );
```
Second question: Does the UI allow the user to create multiple of the same type of automation rules? If so, is it common for "last one wins" to be applied, as seems to happen here?

kevinfodness replied:
Yes, and yes. For example, there might be one behavior if there is a particular category applied, but if there is a specific tag applied, the publisher may want different behavior that is more specific than the generic behavior for the category, so there's the ability to override with "last one wins" logic.

dlh01 reviewed (approved):
🌳 Some idle questions.

kevinfodness committed:
APPLE-179 Code review feedback

- Store original titles in an ID/title map
- Fix failing tests due to this change by using the new get_json_for_post method, which emulates a full push and uses a real post
- Avoids prepending a space if the prepend text is empty

kevinfodness committed:
APPLE-179 phpcs
